srml/babe/src/lib.rs

```diff
@@ -241,6 +248,58 @@ impl<T: Trait> session::ShouldEndSession<T::BlockNumber> for Module<T> {
 	}
 }
 
+// TODO [slashing]: @marcio use this, remove the dead_code annotation.
+/// A BABE equivocation offence report.
+///
+/// When a validator released two or more blocks at the same slot.
+#[allow(dead_code)]
+struct BabeEquivocationOffence<FullIdentification> {
+	/// A babe slot number in which this incident happened.
+	slot: u64,
+	/// The session index in which the incident happened.
+	session_index: SessionIndex,
+	/// The size of the validator set at the time of the offence.
+	validator_set_count: u32,
+	/// The authority which produced the equivocation.
+	offender: FullIdentification,
+}
+
+impl<FullIdentification: Clone> Offence<FullIdentification> for BabeEquivocationOffence<FullIdentification> {
+	const ID: Kind = *b"babe:equivocatio";
+
+	fn offenders(&self) -> Vec<FullIdentification> {
+		vec![self.offender.clone()]
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.session_index
+	}
+
+	fn validator_set_count(&self) -> u32 {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> TimeSlot {
+		self.slot as TimeSlot
+	}
+
+	fn slash_fraction(
+		offenders_count: u32,
+		validator_set_count: u32,
+	) -> Perbill {
+		// the formula is min((3k / n)^2, 1)
+		let x = Perbill::from_rational_approximation(3 * offenders_count, validator_set_count);
+
+		// _ ^ 2
+		// For now, Perbill doesn't support taking the power of it. Until it does, do it manually.
+		// The conversion to u64 is performed to guarantee it fits.
+		// TODO: #3189 should fix this.
+		let x = x.into_parts();
+		let x = ((x as u64 * x as u64) / (1_000_000_000 * 1_000_000_000)) as u32;
```

Comment:
1. this is likely to overflow
2. `1_000_000_000`^2 is only 1 order of magnitude away from `u64::max`. The granularity of fractions here is going to suck with integer division

Reply:
Done in https://github.com/paritytech/substrate/pull/3322/commits/0b57b0dd794ecc8947736c24f4c8514bb0352fb3